### PR TITLE
Use OpenSSL to handle ECDSA signature

### LINKED ISSUE #187: Implement ECDSA signer that uses OpenSSL
On Spomky-Labs/jose#161 @skyzyx commented that he's facing some performance issues with this library in his realtime API, based on @Spomky's response I've did some minor changes on his performance test (to adapt to `4.0.0-alpha1`) and I got this performance comparison (running on my laptop - PHP 7.1):

**it's important to say that this comparison is obviously just a reference and doesn't mean absolutely NOTHING in terms of which lib is better or not (specially because the implementations are different and @Spomky's lib supports JWK - that requires an additional complexity)**

**Signature**

Algorithm | Spomky-Labs/jose | lcobucci/jwt 
---------- | -------------------| ----------------
none         | 0.007670 ms/op    | 0.001469  ms/op
HS256      | 0.023170 ms/op    | 0.007911 ms/op
HS384      | 0.021369 ms/op    | 0.007732 ms/op
HS512       | 0.021610 ms/op    | 0.006850 ms/op
RS256       | 19.724061 ms/op  | 1.402881 ms/op
RS384       | 19.679730 ms/op | 1.385570 ms/op
RS512        | 19.765210 ms/op | 1.321239 ms/op
ES256       | 1.326001 ms/op    | 154.087379 ms/op
ES384       | 2.366140 ms/op   | 146.823630 ms/op
ES512        | 2.271872 ms/op    | 145.546641 ms/op

**Verification**

Algorithm | Spomky-Labs/jose | lcobucci/jwt 
---------- | -------------------| ----------------
none         | 0.007741  ms/op    |0.001271  ms/op
HS256     | 0.023739 ms/op     |0.014479 ms/op
HS384     | 0.023310 ms/op     | 0.008581 ms/op
HS512      | 0.023758 ms/op    | 0.008640 ms/op
RS256      | 17.913599 ms/op   | 0.065980 ms/op
RS384      | 18.093328 ms/op  | 0.068090 ms/op
RS512      | 17.817550 ms/op    | 0.067239 ms/op
ES256      | 1.715958 ms/op      | 295.684772 ms/op
ES384     | 3.032808 ms/op     | 283.335109 ms/op
ES512      | 3.423491 ms/op     | 294.083581 ms/op

It's really clear that ECDSA is far from being good (is quite depressing to be honest) in terms of performance and the big difference is that in @Spomky's implementation that algorithm uses primarily OpenSSL and phpecc as fallback when the ECDSA is not supported in server.

I'm still thinking if I want to have the same fallback idea or to provide two different implementations but I'm more inclined to go with the latter.

Thanks @ivanrosolen for pointing out the issue to me and @Spomky for providing a such complete implementation for JOSE 🍻 

@skyzyx would you mind sharing your issues with my lib here as well?

### PR DESCRIPTION
**Pull request by lcobucci**
I finally found some energy to finish this, there're things that could/should be improved though.

Fixes #187

### CODE REVIEW

File: test/performance/SignerBench.php [hunk 1]
```diff
@@ -5,10 +5,12 @@
 
 use Lcobucci\JWT\Signer\Key;
 use PhpBench\Benchmark\Metadata\Annotations\BeforeMethods;
+use PhpBench\Benchmark\Metadata\Annotations\Iterations;
 use PhpBench\Benchmark\Metadata\Annotations\Revs;
 
 /**
  * @BeforeMethods({"init"})
+ * @Iterations(5)
```
Comment (lcobucci) [inline test/performance/SignerBench.php]: Unrelated, I know... but whateva

File: src/Signer/OpenSSL.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+use InvalidArgumentException;
+use Lcobucci\JWT\Signer;
+use function openssl_error_string;
+use function openssl_pkey_get_details;
+use function openssl_pkey_get_private;
+use function openssl_pkey_get_public;
+use function openssl_sign;
+use function openssl_verify;
+
+abstract class OpenSSL implements Signer
```
Comment (lcobucci) [inline src/Signer/OpenSSL.php]: This introduces another level to the hierarchy, adding more 💩... I didn't want to do a bigger change, though

File: test/unit/Signer/RsaTest.php [hunk 3]
```diff
@@ -6,8 +6,8 @@
 use Lcobucci\JWT\Keys;
 use PHPUnit\Framework\TestCase;
 use const OPENSSL_ALGO_SHA256;
-use function openssl_get_privatekey;
-use function openssl_get_publickey;
+use function openssl_pkey_get_private;
```
Comment (lcobucci) [inline test/unit/Signer/RsaTest.php]: Unrelated to the change itself, just decided to not use aliases

File: test/unit/Signer/RsaTest.php [hunk 4]
```diff
@@ -18,10 +18,10 @@
     /**
      * @test
      *
-     * @requires extension openssl
```
Comment (lcobucci) [inline test/unit/Signer/RsaTest.php]: The project depends on `openssl` there's no point in having this here

File: test/unit/Signer/RsaTest.php [hunk 5]
```diff
@@ -18,10 +18,10 @@
...
-     *
      * @covers \Lcobucci\JWT\Signer\Rsa::sign
      * @covers \Lcobucci\JWT\Signer\Rsa::validateKey
+     * @covers \Lcobucci\JWT\Signer\Rsa::getKeyType
+     * @covers \Lcobucci\JWT\Signer\OpenSSL
```
Comment (lcobucci) [inline test/unit/Signer/RsaTest.php]: This is weird, I know... we should simply use OpenSSL and not cover it here. Will open an issue to address this.

File: src/Signer/Ecdsa/PointsManipulator.php [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<?php
+declare(strict_types=1);
+
+namespace Lcobucci\JWT\Signer\Ecdsa;
+
+/**
+ * @internal
+ */
+interface PointsManipulator
```
Comment (lcobucci) [inline src/Signer/Ecdsa/PointsManipulator.php]: Naming is hard, you know!?

File: src/Signer/Ecdsa/PointsManipulator.php [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public function fromEcPoint(string $signature, int $length): string;
```
Comment (lcobucci) [inline src/Signer/Ecdsa/PointsManipulator.php]: This is also bit weird in terms of naming, open to suggestions

File: composer.json [hunk 8]
```diff
@@ -18,14 +18,16 @@
     ],
     "require": {
         "php": "^7.1",
+        "ext-gmp": "*",
```
Comment (Ocramius) [inline composer.json]: Argh, this is kinda heavy.

File: src/Signer/Ecdsa/Asn1.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+use function hex2bin;
+use function mb_strlen;
+use function mb_substr;
+use function str_pad;
+
+final class Asn1 implements PointsManipulator
+{
+    public function fromEcPoint(string $signature, int $length): string
+    {
+        $sequence = ASNObject::fromBinary($signature);
```
Comment (Ocramius) [inline src/Signer/Ecdsa/Asn1.php]: Does `ASN` stand for `ASshole Number`? O_o (ok, I should read up on this stuff)

File: src/Signer/Ecdsa.php [hunk 10]
```diff
@@ -3,63 +3,51 @@
...
      * {@inheritdoc}
      */
     final public function sign(string $payload, Key $key): string
     {
-        return $this->adapter->createHash(
-            $this->keyParser->getPrivateKey($key),
-            $this->adapter->createSigningHash($payload, $this->getAlgorithm()),
-            $this->getAlgorithm()
-        );
+        $signature = $this->createSignature($key->getContent(), $key->getPassphrase(), $payload);
```
Comment (Ocramius) [inline src/Signer/Ecdsa.php]: Redundant variable

File: src/Signer/Ecdsa.php [hunk 11]
```diff
@@ -3,63 +3,51 @@
...
      */
     final public function verify(string $expected, string $payload, Key $key): bool
     {
-        return $this->adapter->verifyHash(
-            $expected,
-            $this->keyParser->getPublicKey($key),
-            $this->adapter->createSigningHash($payload, $this->getAlgorithm()),
-            $this->getAlgorithm()
-        );
+        $signature = $this->manipulator->toEcPoint($expected, $this->getKeyLength());
```
Comment (Ocramius) [inline src/Signer/Ecdsa.php]: Redundant variable

File: src/Signer/Ecdsa/Asn1.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    public function toEcPoint(string $points, int $length): string
+    {
+        $points = bin2hex($points);
+
+        if (mb_strlen($points, '8bit') !== 2 * $length) {
+            throw new InvalidArgumentException('The length of given value is different than expected');
+        }
+
+        $pointR = mb_substr($points, 0, $length, '8bit');
```
Comment (Ocramius) [inline src/Signer/Ecdsa/Asn1.php]: Any kind of naming that would make this easier to understand, or is this from a reference implementation? If it is so, then let's please `@link` it

File: src/Signer/Ecdsa/EccAdapter.php [hunk 13]
```diff
@@ -1,118 +0,0 @@
...
-use Mdanter\Ecc\Crypto\Signature\Signer;
-use Mdanter\Ecc\Curves\NistCurve;
-use Mdanter\Ecc\EccFactory;
-use Mdanter\Ecc\Math\GmpMathInterface;
-use Mdanter\Ecc\Primitives\GeneratorPoint;
-use Mdanter\Ecc\Random\RandomGeneratorFactory;
-use Mdanter\Ecc\Random\RandomNumberGeneratorInterface;
-use function array_key_exists;
-
-class EccAdapter
```
Comment (Ocramius) [inline src/Signer/Ecdsa/EccAdapter.php]: this BC break requires documentation

File: src/Signer/Ecdsa/KeyParser.php [hunk 14]
```diff
@@ -1,93 +0,0 @@
...
-use Mdanter\Ecc\Serializer\PrivateKey\PrivateKeySerializerInterface;
-use Mdanter\Ecc\Serializer\PublicKey\DerPublicKeySerializer;
-use Mdanter\Ecc\Serializer\PublicKey\PemPublicKeySerializer;
-use Mdanter\Ecc\Serializer\PublicKey\PublicKeySerializerInterface;
-use const PHP_EOL;
-use function count;
-use function preg_match;
-use function str_replace;
-
-class KeyParser
```
Comment (Ocramius) [inline src/Signer/Ecdsa/KeyParser.php]: This BC break requires documentation

File: src/Signer/Ecdsa/PointsManipulator.php (hunk 6 above)
Comment (Ocramius) [inline src/Signer/Ecdsa/PointsManipulator.php]: Can you please still describe the interface? Unclear what it does, or what the naming means. If it comes from somewhere, `@link` is vital.

File: src/Signer/Ecdsa/Sha256.php [hunk 15]
```diff
@@ -18,8 +19,13 @@ public function getAlgorithmId(): string
     /**
      * {@inheritdoc}
      */
-    public function getAlgorithm(): string
+    public function getAlgorithm(): int
```
Comment (Ocramius) [inline src/Signer/Ecdsa/Sha256.php]: BC break to be documented

File: src/Signer/Ecdsa/Sha384.php [hunk 16]
```diff
@@ -18,8 +19,13 @@ public function getAlgorithmId(): string
     /**
      * {@inheritdoc}
      */
-    public function getAlgorithm(): string
+    public function getAlgorithm(): int
```
Comment (Ocramius) [inline src/Signer/Ecdsa/Sha384.php]: BC break to be documented

File: src/Signer/Ecdsa/Sha512.php [hunk 17]
```diff
@@ -18,8 +19,13 @@ public function getAlgorithmId(): string
     /**
      * {@inheritdoc}
      */
-    public function getAlgorithm(): string
+    public function getAlgorithm(): int
```
Comment (Ocramius) [inline src/Signer/Ecdsa/Sha512.php]: BC break to be documented

File: src/Signer/Ecdsa/SignatureSerializer.php [hunk 18]
```diff
@@ -1,72 +0,0 @@
...
-use Mdanter\Ecc\Math\GmpMathInterface;
-use const STR_PAD_LEFT;
-use function gmp_init;
-use function pack;
-use function sprintf;
-use function str_pad;
-use function str_split;
-use function unpack;
-
-class SignatureSerializer
```
Comment (Ocramius) [inline src/Signer/Ecdsa/SignatureSerializer.php]: BC break to be documented

File: src/Signer/OpenSSL.php [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        $details = openssl_pkey_get_details($key);
+
+        if (! isset($details['key']) || $details['type'] !== $this->getKeyType()) {
+            throw new InvalidArgumentException('This key is not compatible with this signer');
+        }
+    }
+
+    abstract public function getAlgorithm(): int;
```
Comment (Ocramius) [inline src/Signer/OpenSSL.php]: Requires better documentation

File: src/Signer/OpenSSL.php [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+    abstract public function getKeyType(): int;
```
Comment (Ocramius) [inline src/Signer/OpenSSL.php]: Requires better documentation

File: src/Signer/Ecdsa/Asn1.php [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+use function assert;
+use function bin2hex;
+use function gmp_init;
+use function gmp_strval;
+use function hex2bin;
+use function mb_strlen;
+use function mb_substr;
+use function str_pad;
+
+final class Asn1 implements PointsManipulator
```
Comment (Ocramius) [inline src/Signer/Ecdsa/Asn1.php]: This implementation doesn't appear to be tested

File: composer.json (hunk 8 above)
Comment (lcobucci) [reply composer.json]: Yeap, it reinforces my wish to split the lib in different small ones.

File: src/Signer/Ecdsa/Asn1.php (hunk 21 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/Asn1.php]: It's being tested through the ecdsa test, which is not ideal (just like the OpenSSL)

File: src/Signer/Ecdsa/Asn1.php (hunk 12 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/Asn1.php]: This is essentially the logic [required by JWA](https://tools.ietf.org/html/rfc7518#page-90). I can add the link there.

File: src/Signer/Ecdsa/EccAdapter.php (hunk 13 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/EccAdapter.php]: This class was always considered as internal, but we were not using `@internal` yet. I'm not sure if this really need to be stated as BC-break :(

File: src/Signer/Ecdsa/KeyParser.php (hunk 14 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/KeyParser.php]: This class was always considered as internal, but we were not using `@internal` yet. I'm not sure if this really need to be stated as BC-break :(

File: src/Signer/Ecdsa/Sha256.php (hunk 15 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/Sha256.php]: This method was always considered as internal, but we were not using `@internal` yet. I'm not sure if this really need to be stated as BC-break :(

File: src/Signer/Ecdsa/Sha384.php (hunk 16 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/Sha384.php]: This method was always considered as internal, but we were not using `@internal` yet. I'm not sure if this really need to be stated as BC-break :(

File: src/Signer/Ecdsa/Sha512.php (hunk 17 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/Sha512.php]: This method was always considered as internal, but we were not using `@internal` yet. I'm not sure if this really need to be stated as BC-break :(

File: src/Signer/Ecdsa/SignatureSerializer.php (hunk 18 above)
Comment (lcobucci) [reply src/Signer/Ecdsa/SignatureSerializer.php]: This method was always considered as internal, but we were not using `@internal` yet. I'm not sure if this really need to be stated as BC-break :(

**Pull request merged**